xc7/utils/prjxray_edge_library.py

```diff
@@ -0,0 +1,1834 @@
+        for src_graph_node_pkey, switch_pkey, dest_graph_node_pkey in sink_connector.connect_at(
+                pip_pkey=pip_pkey, src_wire_pkey=sink_wire_pkey,
+                dest_wire_pkey=src_wire_pkey, loc=loc,
+                other_connector=src_connector):
+            yield (
+                src_graph_node_pkey, dest_graph_node_pkey, switch_pkey,
+                phy_tile_pkey, pip_pkey, True
+            )
+
+    # Make additional connections to constant network if the sink needs it.
+    for constant_src in yield_ties_to_wire(pip.net_to):
+        for src_graph_node_pkey, switch_pkey, dest_graph_node_pkey in const_connectors[
+                constant_src].connect_at(switch_pkey=delayless_switch_pkey,
+                                         loc=loc,
+                                         other_connector=sink_connector):
+            yield (
+                src_graph_node_pkey, dest_graph_node_pkey, switch_pkey,
+                phy_tile_pkey, None, False
+            )
+
+
+def mark_track_liveness(conn, input_only_nodes, output_only_nodes):
+    """ Checks tracks for liveness.
+
+    Iterates over all graph nodes that are routing tracks and determines if
+    at least one graph edge originates from or two the track.
+
+    Args:
+        conn (sqlite3.Connection): Connection database
+
+    """
+
+    alive_tracks = set()
+    c = conn.cursor()
+    write_cur = conn.cursor()
+    for graph_node_pkey, node_pkey, track_pkey in c.execute("""
+SELECT
+  pkey,
+  node_pkey,
+  track_pkey
+FROM
+  graph_node
+WHERE
+  track_pkey IS NOT NULL;"""):
+        if track_pkey in alive_tracks:
+            continue
+
+        if node_pkey in input_only_nodes or node_pkey in output_only_nodes:
+            alive_tracks.add(track_pkey)
+            continue
+
+        write_cur.execute(
+            """SELECT count(switch_pkey) FROM graph_edge WHERE
+            src_graph_node_pkey = ?;""", (graph_node_pkey, )
+        )
+        src_count = write_cur.fetchone()[0]
+
+        write_cur.execute(
+            """SELECT count(switch_pkey) FROM graph_edge WHERE
+            dest_graph_node_pkey = ?;""", (graph_node_pkey, )
+        )
+        sink_count = write_cur.fetchone()[0]
+
+        if src_count > 0 and sink_count > 0:
+            alive_tracks.add(track_pkey)
+
+    c.execute("SELECT count(pkey) FROM track;")
+    track_count = c.fetchone()[0]
+    print(
+        "{} Alive tracks {} / {}".format(
+            now(), len(alive_tracks), track_count
+        )
+    )
+
+    write_cur.execute("""BEGIN EXCLUSIVE TRANSACTION;""")
+    for (track_pkey, ) in c.execute("""SELECT pkey FROM track;"""):
+        write_cur.execute(
+            "UPDATE track SET alive = ? WHERE pkey = ?;",
+            (track_pkey in alive_tracks, track_pkey)
+        )
+    write_cur.execute("""COMMIT TRANSACTION;""")
+
+    print('{} Track aliveness committed'.format(now()))
+
+    write_cur.execute("""BEGIN EXCLUSIVE TRANSACTION;""")
```

Comment:
Does each line need to be executed separately?

Reply:
Yes, https://docs.python.org/2/library/sqlite3.html#sqlite3.Cursor.execute